gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java

```diff
@@ -4049,8 +4061,25 @@ public boolean isSnapshot() {
   }
 
   @Override
-  public void recordVersionForSnapshot(Object member, long version, Region region ) {
+  public void recordVersionForSnapshot(Object member, long version, Region region) {
     queue.add(new VersionInformation(member, version, region));
+    Boolean wasPresent = writeRegions.putIfAbsent(region, true);
+    // Handle both row buffer and column tables together , when column  table
+    // insert is transactional , both of them can have data inserted from a transaction
+    // and it might lead to deadlock.
+    if (wasPresent == null) {
+      if (region instanceof BucketRegion) {
+        BucketRegion br = (BucketRegion)region;
+        BucketRegion companionRegion = br.companionRegion();
+        if (companionRegion != null) {
```

Comment:
Shadow region's companion region is row buffer & row buffer's companion region is shadow region. We need to take a read lock at one time as GII might lock one and wait for other while transaction has locked the previous one. That will cause a deadlock. 
Direct transfer to cached batch region won't be problematic. Only we will be taking a read lock on row buffer unnecessarily.